### PR TITLE
BoundedSourceQueue API

### PR DESCRIPTION
**Pull request by seglo**
Based on @jrudolph's FastDroppingQueue implementation in #29574 

A new queue implementation that drops new elements immediately when the buffer is full. Does not use async callbacks like `Source.queue` with `OverflowStrategy.dropNew`, which can still result in OOM errors (#25798).

Other "drop" scenarios (`dropTail`, `dropHead`) may be supported by this implementation if their use can be justified.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:915]: Sorry, I haven't been able to review this PR yet, but I think `boundedQueue` would be a better name.

Even better would be to keep it as `Source.queue` overload, but maybe we are using same parameters?

Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:915]: > I think boundedQueue would be a better name

SGTM

> Even better would be to keep it as `Source.queue` overload, but maybe we are using same parameters?

We could use the overload `def queue(bufferSize: Int): Source[T, BoundedQueueSource[T]]`.  It will have a different return type than the other overloads, as well as a different implementation.  

We could discuss deprecating the other overloads and supporting the back-pressure (and maybe the fail) `OverflowStrategy` as separate `Source` factory methods.

Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: I renamed the impl. to `BoundedQueue`.

I'm now using the `Source.queue(bufferSize: Int)` overload for the new impl and swapping the impl. when the `OverflowStrategy.dropNew` strategy is used in the other overloads.

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (raboof) [inline akka-docs/src/main/paradox/stream/operators/Source/queue.md:29]:
```suggestion
When the buffer is full, the `BoundedQueueSource` will not accept further elements.
```

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/QueueSourceSpec.scala
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/QueueSourceSpec.scala:263]: This is a pretty intricate way of introducing the cancellation... does the `sub.cancel()` no longer work?

**Review by raboof — COMMENTED** [review]
Looking good overall!

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:798]: Are we sure that we want to change the queue implementation for DropNew? It's probably much better, but the change in behavior could impact users. I think we should consider deprecating this version of `Source.queue` instead and promote the new method instead. The deprecation could come in a separate PR (possibly later release)  since we probably want to provide a good alternative replacement for at least `Backpressure` strategy.

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:29]: :+1:

File: akka-docs/src/test/java/jdocs/stream/IntegrationDocTest.java
Comment (seglo) [reply akka-docs/src/test/java/jdocs/stream/IntegrationDocTest.java:786]: I didn't put much thought into this example. I just copied the other one and modified it to fit the API. I'll think about it some more, but will just use an array for now as you suggested.

File: akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala
Comment (seglo) [reply akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala:510]: :+1:

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/QueueSourceSpec.scala
Comment (seglo) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/QueueSourceSpec.scala:263]: It works, although it returned a different stream "failure". I'm a little confused about this TBH since I'm not sure why a failure should be returned if a completion occurs.

Comment (raboof) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/QueueSourceSpec.scala:263]: > I'm a little confused about this TBH since I'm not sure why a failure should be returned if a completion occurs.

That's a good point: if downstream is no longer interested in more elements, should that look like a 'failure' to upstream or should those elements just disappear into the void? I can see pros and cons for either interpretation...

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:9]: :+1:

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: I'll revert this change.

In the future I suggest we deprecate these two overloads altogether (with a backpressure and "fail" strategy alternative).  We can't deprecate `OverflowStrategies.DropNew` because it's used in `Source.actorRef` as well.

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: and we might not need a fail strategy because that can be a decision for the caller (and implemented together with a KillSwitch)

**Review by patriknw — APPROVED** [review]
LGTM, something minor...

Before merging we should manually squash this into 2 commits (one by Johannes, and one by Sean) and NOT use squash button.

File: akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala
Comment (seglo) [reply akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala:504]: :+1:

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:705]: Good catch. I updated other `QueueOfferResult` references in the `javadsl.Source` docs too.

Should we refer to `QueueClosed` as `QueueOfferResult.QueueClosed` or `QueueOfferResult.QueueClosed$` ?

**Comment by seglo** [comment]
> Before merging we should manually squash this into 2 commits (one by Johannes, and one by Sean) and NOT use squash button.

:+1: 

I squashed the commits, but I had to rebase onto master when creating the PR because #29574 targeted `release-2.5`, therefore Johannes' commits were rewritten (but author id is preserved).  LMK if this is acceptable @patriknw, @jrudolph.

**Comment by patriknw** [comment]
preserving the author was the point, so good

File: akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]:
```
 * Not for user extension
 */
@DoNotInherit
```

File: akka-stream/src/main/scala/akka/stream/QueueOfferResult.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/QueueOfferResult.scala:8]: Add some `@DoNotInherit` for these two as well, sealed doesn't help Java users.

File: akka-stream/src/main/scala/akka/stream/impl/BoundedQueueSource.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/BoundedQueueSource.scala:33]: As noted above, does it really make sense that it can be 0?

**Review by johanandren — COMMENTED** [review]
Looking good, some small things noted.

File: akka-docs/src/main/paradox/stream/operators/index.md
Comment (jrudolph) [inline akka-docs/src/main/paradox/stream/operators/index.md:38]: To keep the queues and attributes in the same order:

```suggestion
|Source|<a name="queue"></a>@ref[queue](Source/queue.md)|Materialize a `BoundedSourceQueue` or `SourceQueue` onto which elements can be pushed synchronously or asynchronously for emitting from the source.|
```

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (jrudolph) [inline akka-docs/src/main/paradox/stream/operators/Source/queue.md:3]: Here's a suggestion of giving an overview about the options right away (I found the distinction between `synchronous` and `asynchronous` not really descriptive as a user coming to this page):

```suggestion
Materialize a `BoundedSourceQueue` or `SourceQueue` onto which elements can be pushed for emitting from the source.

The `BoundedSourceQueue` is an optimized variant of the `SourceQueue` with `OverflowStrategy.dropNew`. The `BoundedSourceQueue` will give immediate, synchronous feedback whether an element was accepted or not and is therefore recommended for situations where overload and dropping elements is expected and needs to be handled quickly.

In contrast, the `SourceQueue` offers more variety of `OverflowStrategies` but feedback is only asynchronously provided through a @scala[`Future`]@java[`CompletionStage`] value. In cases where elements need to be discarded quickly at times of overload to avoid out-of-memory situations, delivering feedback asynchronously can itself become a problem. This happens if elements come in faster than the feedback can be delivered in which case the feedback mechanism itself is part of the reason that an out-of-memory situation arises.

In summary, prefer `BoundedSourceQueue` over `SourceQueue` with `OverflowStrategy.dropNew` especially in high-load scenarios. Use `SourceQueue` if you need one of the other `OverflowStrategies`.
```

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (jrudolph) [inline akka-docs/src/main/paradox/stream/operators/Source/queue.md:7]: Maybe let's use the name of the materialized queue as the identifier for the two variants?

```suggestion
## Signature (`BoundedSourceQueue`)
```

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (jrudolph) [inline akka-docs/src/main/paradox/stream/operators/Source/queue.md:16]:
```suggestion
The return value of `BoundedQueueSource.offer()` immediately returns a `QueueOfferResult` (as opposed to an asynchronous value as with `SourceQueue).
```

File: akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]: Why not `BoundedSourceQueue`? After all, it's not a source but a queue and that would also be consistent with the existing naming? (I just realized that I confused those also above in my docs suggestions).

Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]: Btw. I think it would be fine if a `BoundedQueueSource(Stage)` would provide a `BoundedSourceQueue`. After all these are two opposite ends of looking at the component.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:700]: "queue is completed" - Why would that happen if the queue is completed? What does that mean?

"stream completes" - What does that mean?

E.g. clearer like this:

```suggestion
   * queue is failed through the materialized `BoundedQueueSource` or the source is cancelled by the downstream.
```

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:705]:
```suggestion
   * `QueueOfferResult.enqueued()`     element was added to buffer (but may still be discarded later when the queue is failed or cancelled)
```

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:711]:
```suggestion
   * @param bufferSize size of the buffer in number of elements
```

or maybe just leave it away

**Review by jrudolph — COMMENTED** [review]
Thanks for doing this work, @seglo! I mostly looked at the documentation. I guess my biggest non-cosmetic suggestion is to rename `BoundedQueueSource` to `BoundedSourceQueue`...

**Comment by jrudolph** [comment]
> LMK if this is acceptable @patriknw, @jrudolph.

Of course, thanks a lot for bringing that over the finish line.

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:17]: :+1:

File: akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala
Comment (seglo) [reply akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala:499]: This is the result of `scalafmt` rules, but I'll add a formatting exception around it.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/BoundedQueueSourceSpec.scala
Comment (seglo) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/BoundedQueueSourceSpec.scala:68]: That makes sense to me.  I set it to `0` because the `AbstractBoundedNodeQueue` actually supports a buffer size of 0, but it's not clear to me why.

File: akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]: I agree it doesn't make sense for a user to implement this trait. I followed the convention from the `javadsl|scaladsl.Queue` trait which doesn't have `@DoNotInherit`.

Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]: I'll add `@DoNotInherit` to `Queue` as well.

Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]: Looks like all of the mat types don't include this attribute.  I'll just apply it to `BoundedQueueSource`.

File: akka-stream/src/main/scala/akka/stream/QueueOfferResult.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/QueueOfferResult.scala:8]: :+1:

**Comment by seglo** [comment]
> I guess my biggest non-cosmetic suggestion is to rename `BoundedQueueSource` to `BoundedSourceQueue`

I originally used that convention, but I changed it to what it is now for some reason.  I'll make it `BoundedSourceQueue`.

:+1:

File: akka-docs/src/main/paradox/stream/operators/index.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/index.md:38]: :+1:

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:3]: I initially had more content above the fold like you suggested, but a docs post-processing step didn't like it.  I'll take another look to see if I can make that work.

Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:3]: It looks like a rule was explicitly created to keep the description one line (to keep it short).

https://github.com/akka/akka/blob/master/project/StreamOperatorsIndexGenerator.scala#L244-L248

I like your description though, so I'll try to make it work with the rule intact.

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:7]: :+1:

File: akka-docs/src/main/paradox/stream/operators/Source/queue.md
Comment (seglo) [reply akka-docs/src/main/paradox/stream/operators/Source/queue.md:16]: :+1:

File: akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/BoundedQueueSource.scala:10]: I renamed `BoundedQueueSource` to `BoundedSourceQueue`.  I changed the stage too just to make it consistent.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:700]: SGTM

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:701]: :+1:

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:705]: :+1:

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:711]: :+1:

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/BoundedQueueSourceSpec.scala
Comment (seglo) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/BoundedQueueSourceSpec.scala:68]: I updated the `bufferSize` constraint to be `> 0`

File: akka-stream/src/main/scala/akka/stream/impl/BoundedQueueSource.scala
Comment (seglo) [reply akka-stream/src/main/scala/akka/stream/impl/BoundedQueueSource.scala]: Updated to be `> 0`

**Comment by seglo** [comment]
@johanandren @jrudolph Thanks for the thorough review. I accommodated almost all of your feedback. I'll be away next week, but feel free to push commits if necessary to have this completed this sprint.

**Review by johanandren — APPROVED** [review]

**Comment by johanandren** [comment]
Given that you are away next week, and @jrudolph is away today, let's merge this and follow up with fixes if J has any more feedback when back next week.

**Comment by seglo** [comment]
Sounds good.  Thanks @johanandren :+1:

**Pull request merged**
